### PR TITLE
Fix Number::StepIterator overflow

### PR DESCRIPTION
**Pull request by bcardiff**
Before this PR the `Number::StepIterator#next` could raise because of `limit - @current` 
Fixes a regression introduced in #10130

```
Failures:

  1) Number #step whole range iterates UInt8 downwards

       Arithmetic overflow (OverflowError)
         from src/number.cr:249:21 in 'next'
         from src/iterator.cr:488:15 in '->'
         from src/primitives.cr:255:3 in 'internal_run'
         from src/spec/example.cr:33:16 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:330:7 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:330:7 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:330:7 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:147:7 in 'run'
         from src/spec/dsl.cr:274:7 in '->'
         from src/primitives.cr:255:3 in 'run'
         from src/crystal/main.cr:45:14 in 'main'
         from src/crystal/main.cr:119:3 in 'main'
       

  2) Number #step whole range iterates Int8 upwards

       Arithmetic overflow (OverflowError)
         from src/number.cr:249:21 in 'next'
         from src/iterator.cr:488:15 in '->'
         from src/primitives.cr:255:3 in 'internal_run'
         from src/spec/example.cr:33:16 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:330:7 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:330:7 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:330:7 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:147:7 in 'run'
         from src/spec/dsl.cr:274:7 in '->'
         from src/primitives.cr:255:3 in 'run'
         from src/crystal/main.cr:45:14 in 'main'
         from src/crystal/main.cr:119:3 in 'main'
       

  3) Number #step whole range iterates Int8 downwards

       Arithmetic overflow (OverflowError)
         from src/number.cr:249:21 in 'next'
         from src/iterator.cr:488:15 in '->'
         from src/primitives.cr:255:3 in 'internal_run'
         from src/spec/example.cr:33:16 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:330:7 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:330:7 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:330:7 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:147:7 in 'run'
         from src/spec/dsl.cr:274:7 in '->'
         from src/primitives.cr:255:3 in 'run'
         from src/crystal/main.cr:45:14 in 'main'
         from src/crystal/main.cr:119:3 in 'main'
       

  4) Number #step whole range iterates Int16 upwards

       Arithmetic overflow (OverflowError)
         from src/number.cr:249:21 in 'next'
         from src/iterator.cr:488:15 in '->'
         from src/primitives.cr:255:3 in 'internal_run'
         from src/spec/example.cr:33:16 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:330:7 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:330:7 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:330:7 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:147:7 in 'run'
         from src/spec/dsl.cr:274:7 in '->'
         from src/primitives.cr:255:3 in 'run'
         from src/crystal/main.cr:45:14 in 'main'
         from src/crystal/main.cr:119:3 in 'main'
       

  5) Number #step whole range iterates Int16 downwards

       Arithmetic overflow (OverflowError)
         from src/number.cr:249:21 in 'next'
         from src/iterator.cr:488:15 in '->'
         from src/primitives.cr:255:3 in 'internal_run'
         from src/spec/example.cr:33:16 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:330:7 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:330:7 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:330:7 in 'run'
         from src/spec/context.cr:18:23 in 'internal_run'
         from src/spec/context.cr:147:7 in 'run'
         from src/spec/dsl.cr:274:7 in '->'
         from src/primitives.cr:255:3 in 'run'
         from src/crystal/main.cr:45:14 in 'main'
         from src/crystal/main.cr:119:3 in 'main'
       

Finished in 53.71 milliseconds
733 examples, 0 failures, 5 errors, 0 pending

Failed examples:

crystal spec spec/std/number_spec.cr:309 # Number #step whole range iterates UInt8 downwards
crystal spec spec/std/number_spec.cr:313 # Number #step whole range iterates Int8 upwards
crystal spec spec/std/number_spec.cr:318 # Number #step whole range iterates Int8 downwards
crystal spec spec/std/number_spec.cr:322 # Number #step whole range iterates Int16 upwards
crystal spec spec/std/number_spec.cr:327 # Number #step whole range iterates Int16 downwards
```

### CODE REVIEW

**Comment by straight-shoota** [comment]
The same overflow happens with the yielding method as well.

The implementation looks very hard to follow. I'll give it a try to find a better solution.

**Comment by straight-shoota** [comment]
Oh damn: It just needs to be `(limit - step) <=> current` instead of `(limit - current) <=> step` 😆

File: spec/std/number_spec.cr [hunk 1]
```diff
@@ -299,6 +299,24 @@ describe "Number" do
...
+        it_iterates "UInt8 upwards", (UInt8::MIN.to_i..UInt8::MAX.to_i).map(&.to_u8), (UInt8::MIN..UInt8::MAX).step(by: 1)
+        it_iterates "UInt8 downwards", (UInt8::MIN.to_i..UInt8::MAX.to_i).map(&.to_u8).reverse, (UInt8::MAX..UInt8::MIN).step(by: -1)
+
+        it { (Int8::MIN..Int8::MAX).each.count { true }.should eq(256) }
+        it_iterates "Int8 upwards", (Int8::MIN.to_i..Int8::MAX.to_i).map(&.to_i8), (Int8::MIN..Int8::MAX).step(by: 1)
+        it_iterates "Int8 downwards", (Int8::MIN.to_i..Int8::MAX.to_i).map(&.to_i8).reverse, (Int8::MAX..Int8::MIN).step(by: -1)
+
+        it { (Int16::MIN..Int16::MAX).each.count { true }.should eq(65536) }
+        it_iterates "Int16 upwards", (Int16::MIN.to_i..Int16::MAX.to_i).map(&.to_i16), (Int16::MIN..Int16::MAX).step(by: 1)
+        it_iterates "Int16 downwards", (Int16::MIN.to_i..Int16::MAX.to_i).map(&.to_i16).reverse, (Int16::MAX..Int16::MIN).step(by: -1)
```
Comment (straight-shoota) [inline spec/std/number_spec.cr:314]: Collecting the entire ranges in arrays isn't super efficient, but not a big deal either. Using `it_iterates` ensures these test are run against both `step` methods.

**Comment by bcardiff** [comment]
Cool LGTM. I will need your approval on the PR to merge your changes once the CI is happy.

**Review by straight-shoota — APPROVED** [review]
